src/spatial_data_format.rs

```diff
@@ -0,0 +1,515 @@
+
+        (self.header.sample_rate, res_vecs)
+    }
+
+    /// Extracts the streams from a [GrapeFile], interpolating or quantizing
+    /// the streams to produce datapoints at the requested sample rate.
+    pub fn streams_with_sample_rate(&self, sample_rate: u64) -> Vec<(GrapeTag, Vec<f32>)> {
+        match sample_rate.cmp(&self.header.sample_rate) {
+            Ordering::Equal => self.streams_native_sample_rate().1,
+            Ordering::Less => self.streams_quantized(sample_rate),
+            Ordering::Greater => self.streams_interpolated(sample_rate),
+        }
+    }
+
+    /// Take a slice of [GrapeTag]s and sample vectors and zip them.
+    fn attach_tags(tags: &[GrapeTag], samples: Vec<Vec<f32>>) -> Vec<(GrapeTag, Vec<f32>)> {
+        assert_eq!(tags.len(), samples.len());
+        tags.iter().cloned().zip(samples).collect()
+    }
+
+    /// Returns a cloned, de-interleaved version of the samples in the file.
+    fn get_raw_streams(&self) -> Vec<Vec<f32>> {
+        let n_streams = self.header.n_streams as usize;
+        if n_streams == 0 {
+            return Vec::new();
+        }
+        let stream_len = self.samples.len() / n_streams;
+        let mut sample_vecs = vec![Vec::with_capacity(stream_len); n_streams];
+
+        for (i, &e) in self.samples.iter().enumerate() {
+            let stream_idx = i % n_streams;
+            sample_vecs[stream_idx].push(e);
+        }
+        sample_vecs
+    }
+
+    /// Extracts the streams from the file, and interpolates data points to
+    /// produce data points at the requrested sample_rate.
+    fn streams_interpolated(&self, sample_rate: u64) -> Vec<(GrapeTag, Vec<f32>)> {
+        debug_assert!(sample_rate > self.header.sample_rate);
+        let samples_per_pt = sample_rate as usize / self.header.sample_rate as usize;
+        let raw_streams = self.get_raw_streams();
+        let interpolated_streams = raw_streams
+            .into_iter()
+            .map(|v| {
+                v.windows(2)
+                    .flat_map(|w| {
+                        let step = (w[1] - w[0]) / samples_per_pt as f32;
+                        (0..samples_per_pt)
+                            .map(|i| w[0] + i as f32 * step)
+                            .collect::<Vec<_>>()
+                    })
+                    .collect()
+            })
+            .collect();
+        Self::attach_tags(&self.header.tags, interpolated_streams)
+    }
+
+    /// Extracts the streams from the file, and quantizes data points to
+    /// produce data points at the requested sample_rate.
+    fn streams_quantized(&self, sample_rate: u64) -> Vec<(GrapeTag, Vec<f32>)> {
+        debug_assert!(sample_rate < self.header.sample_rate);
+        let pts_per_sample = self.header.sample_rate as usize / sample_rate as usize;
+        let raw_streams = self.get_raw_streams();
+        let quantized_streams = raw_streams
+            .into_iter()
+            .map(|v| {
+                v.chunks(pts_per_sample)
```

Comment:
I'm learning so many `Vec` methods today :D

Reply:
`.chunks` and `.windows` are superstars